driven/storage/adapter_v3.go

```diff
@@ -11,51 +14,295 @@ import (
 )
 
+// FindGroupsV3 finds groups with filter
+func (sa *Adapter) FindGroupsV3(clientID string, filter *model.GroupsFilter) ([]model.Group, error) {
+	groupIDs := []string{}
+	var err error
+	var userID *string
+	var memberships model.MembershipCollection
+
+	groupFilter := bson.D{primitive.E{Key: "client_id", Value: clientID}}
+	findOptions := options.Find()
+
+	if filter != nil {
+		groupIDMap := map[string]bool{}
+		if len(filter.GroupIDs) > 0 {
+			for _, groupID := range filter.GroupIDs {
+				groupIDMap[groupID] = true
+			}
+		}
+
+		if filter.MemberUserID == nil && filter.MemberExternalID != nil {
+			var user model.User
+			err := sa.db.users.Find(bson.D{
+				{"client_id", clientID},
+				{"external_id", filter.MemberExternalID},
+			}, &user, nil)
+			if err != nil {
+				userID = &user.ID
+			}
+		}
+		if userID == nil && filter.MemberUserID == nil && filter.MemberID != nil {
+			membership, _ := sa.FindGroupMembershipByID(clientID, *filter.MemberID)
+			if membership != nil {
+				memberships = model.MembershipCollection{
+					Items: []model.GroupMembership{*membership},
+				}
+				userID = &membership.UserID
+				if len(groupIDMap) == 0 || groupIDMap[membership.GroupID] {
+					groupIDs = append(groupIDs, membership.GroupID)
+				}
+			}
+		}
+
+		if filter.MemberUserID != nil {
+			// find group memberships
+			memberships, err = sa.FindGroupMemberships(clientID, model.MembershipFilter{
+				UserID: filter.MemberUserID,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, membership := range memberships.Items {
+				if len(groupIDMap) == 0 || groupIDMap[membership.GroupID] {
+					groupIDs = append(groupIDs, membership.GroupID)
+				}
+			}
+		}
+
+		if filter.MemberExternalID != nil {
+			// find group memberships
+			memberships, err = sa.FindGroupMemberships(clientID, model.MembershipFilter{
+				ExternalID: filter.MemberExternalID,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, membership := range memberships.Items {
+				if len(groupIDMap) == 0 || groupIDMap[membership.GroupID] {
+					groupIDs = append(groupIDs, membership.GroupID)
+				}
+			}
+		}
+
+		if userID != nil && (filter.MemberID == nil && filter.MemberUserID == nil && filter.MemberExternalID == nil) {
+			innerOrFilter := []bson.M{
+				{"_id": bson.M{"$in": groupIDs}},
+				{"privacy": bson.M{"$ne": "private"}},
+			}
+
+			if filter.Title != nil {
+				innerOrFilter = append(innerOrFilter, primitive.M{"$and": []primitive.M{
+					{"title": *filter.Title},
+					{"hidden_for_search": false},
+				}})
+			}
+
+			orFilter := primitive.E{Key: "$or", Value: innerOrFilter}
+
+			groupFilter = append(groupFilter, orFilter)
+		}
+
+		if len(filter.GroupIDs) > 0 {
+			groupFilter = append(groupFilter, primitive.E{Key: "_id", Value: primitive.M{"$in": filter.GroupIDs}})
+		}
+		if len(groupIDs) > 0 && (filter.MemberID != nil || filter.MemberUserID != nil || filter.MemberExternalID != nil) {
```

Comment:
Thanks for this. I would prefer to use groupIDs as inner storage, because the filter is accessible outside of this method and it would be a kind of inconsistency to add more groupIDs. I had some doubts to remove all Member filters, because they produce missconception and easily may be used in a wrong direction. 
@shurwit @roberlander2 Let me know please what do you think?

Reply:
Thanks @mdryankov, I agree with you about not manipulating the filter object by adding new group IDs, so I think the way it is written now is fine.